Make NativeCallableAttribute public.

AaronRobinsonMSFT opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/32462

**TODO:**
- [x] Remove usage overhead by avoiding COM Delegate creation.
- [x] Existing `NativeCallableAttribute` tests pass.
- [x] Write new tests for negative cases.
- [x] Handle exceptions on Windows.
- [x] Properly fail-fast for exceptions on non-Windows.
- [x] Update CrossGen2.

/cc @jkotas @fadimounir @jeffschwMSFT @jkoritzinsky

AaronRobinsonMSFT committed:
Avoid creating a COM Delegate when calling NativeCallableWrapper.

AaronRobinsonMSFT commented:
/cc @Scottj1s

jkotas commented on src/coreclr/src/vm/jithelpers.cpp:
```diff
@@ -5133,6 +5133,23 @@ Thread * __stdcall JIT_InitPInvokeFrame(InlinedCallFrame *pFrame, PTR_VOID StubS
+
+EXTERN_C void JIT_ReversePInvokeExit()
+{
+    Thread* thread = GetThread();
```
It may be a good idea to cache the `Thread*` in the PInvoke frame allocated by the JIT. Fetching it from stack should be faster than fetching it from thread local.

AaronRobinsonMSFT replied:
Sounds good. Does this mean I should bring back the `ReversePInvokeFrame` I proposed?

jkotas replied:
You just need to store a single `Thread*` pointer on the stack for this. The `ReversePInvokeFrame` you have proposed was much bigger than that.

You can wrap it in a struct if you would like for clarity. But it should be just a single `Thread*` pointer.

jkotas commented on src/coreclr/src/vm/jithelpers.cpp:
```diff
@@ -5133,6 +5133,23 @@ Thread * __stdcall JIT_InitPInvokeFrame(InlinedCallFrame *pFrame, PTR_VOID StubS
+    Thread* thread = GetThreadNULLOk();
+    if (thread == NULL)
+        thread = CreateThreadBlockThrow();
+
```
The existing slow implementation has sanity check for bad ReversePInvoke use (look for `ReversePInvokeBadTransition`). It costs like two extra instructions, but I think is very well worth it. It would be nice to preserve it for this fast path.

jkotas commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -12372,6 +12378,9 @@ CorJitResult CallCompileMethodWithSEHWrapper(EEJitManager *jitMgr,
          }
     }
 
+    if (ftn->HasNativeCallableAttribute())
```
This will also need to be duplicated for crossgen2.

AaronRobinsonMSFT replied:
I believe this is already done.

https://github.com/dotnet/runtime/blob/b0d39e472ae1ce615818f6d8bcc78597e3d2eac4/src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs#L2838-L2839

jkotas replied:
Interesting. But I doubt that it is working end-to-end. 

We may also need to add `READYTORUN_HELPER_...` for the two new helpers to make it actually work.

jkotas replied:
Also, crossgen2 should get the same checks for invalid use as the runtime.

AaronRobinsonMSFT replied:
I'm not entirely sure where these checks should go. Can you point out an example of when this check would be possible?

jkotas replied:
I am not sure either. @MichalStrehovsky or @trylek should have an opinion.

MichalStrehovsky replied:
> Also, crossgen2 should get the same checks for invalid use as the runtime.

I wonder if it would be better to place the validation in `getCallInfo` in both JitInterface.cpp and Crossgen2.

`getFunctionFixedEntrypoint` is usually not called when crossgenning:

```cpp
#ifdef FEATURE_READYTORUN_COMPILER
        if (tree->AsFptrVal()->gtEntryPoint.addr != nullptr)
        {
            addrInfo = tree->AsFptrVal()->gtEntryPoint;
        }
        else
#endif
        {
            info.compCompHnd->getFunctionFixedEntryPoint(tree->AsFptrVal()->gtFptrMethod, &addrInfo);
        }
```

AaronRobinsonMSFT replied:
@MichalStrehovsky Would you expect the `getCallInfo` to throw an exception/fail? I am unclear what information I would pass back in the `getCallInfo` call. Something like `invalidLdftnCall`? Who would be expected to respond to that?

jkotas replied:
> Would you expect the getCallInfo to throw an exception/fail

Yes

MichalStrehovsky replied:
We should be able to throw from there. E.g. here we already do:

https://github.com/dotnet/runtime/blob/71deb1eea5c1f6866833f06ee5f295106b6e24b6/src/coreclr/src/vm/jitinterface.cpp#L5083-L5089

jkotas commented on src/coreclr/src/vm/prestub.cpp:
```diff
@@ -1867,14 +1897,50 @@ extern "C" PCODE STDCALL PreStubWorker(TransitionBlock * pTransitionBlock, Metho
+// per methoddesc. In addition to installing the new code, this function
+// returns a pointer to the new code for the prestub's convenience.
+//=============================================================================
+extern "C" PCODE STDCALL PreStubWorker(TransitionBlock* pTransitionBlock, MethodDesc* pMD)
```
It would be nice to avoid introducing extra stack frame for the common PreStubWorker.

AaronRobinsonMSFT committed:
Create fast P/Invoke frame.
Reduce indirection during typical PreStub operation (i.e. during COOP).
Verify reverse PInvoke transition is valid.

jkotas commented on src/coreclr/src/vm/prestub.cpp:
```diff
@@ -1767,113 +1767,155 @@ extern "C" MethodDesc * STDCALL PreStubGetMethodDescForCompactEntryPoint (PCODE
-    // Running the PreStubWorker on a method causes us to access its MethodTable
-    g_IBCLogger.LogMethodDescAccess(pMD);
+    // Check what GC mode we are running under.
+    if (!CURRENT_THREAD->PreemptiveGCDisabled())
```
`CURRENT_THREAD` can be null here

AaronRobinsonMSFT replied:
Hmmm. Then that is a long standing bug. The `CURRENT_THREAD` was used in the `pPFrame->Push(CURRENT_THREAD);` from the original code and that dereferences the thread in https://github.com/dotnet/runtime/blob/cf4cc83d44f3ebd1807b6db0ad9e546bdce218b7/src/coreclr/src/vm/frames.cpp#L386-L398

Should I convert this to the same approach as `JIT_ReversePInvokeEnter`? Which I guess at this point would mean the thread would be created here and that call could potentially be asserted instead.

jkotas replied:
It can be NULL only for the reverse PInvoke case that you are introducing. I would change this condition to `if (CURRENT_THREAD == NULL || !CURRENT_THREAD->PreemptiveGCDisabled())` and deal with creating the thread in `PreStubWorker_Preemptive`.

AaronRobinsonMSFT replied:
Oh right. The PREEMPT mode is new... duh. That makes sense.

jkotas commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -9200,20 +9200,25 @@ void CEEInfo::getFunctionFixedEntryPoint(CORINFO_METHOD_HANDLE   ftn,
+    // adheres to the limitations.
     if (pMD->HasNativeCallableAttribute())
     {
-        pResult->addr = (void*)COMDelegate::ConvertToCallback(pMD);
```
It may be a good idea to keep `COMDelegate::ConvertToCall` path for Windows x86, in the first iteration at least. I suspect that making this work for `Windows x86` is going to have a bunch of unique problems.

AaronRobinsonMSFT replied:
Okay.

AaronRobinsonMSFT replied:
Oh right... Windows - boo.

AaronRobinsonMSFT committed:
Properly handle Preemptive thread in PreStubWorker.

AaronRobinsonMSFT committed:
Block fast NativeCallableAttribute on the X86 platform.

AaronRobinsonMSFT committed:
Limit the exclusion for fast NativeCallable to only Win-x86.

AaronRobinsonMSFT committed:
Add Reverse P/Invoke JIT helpers to CrossGen2.

AaronRobinsonMSFT commented:
/cc @MichalStrehovsky for CrossGen2

jkotas commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -9202,18 +9202,40 @@ void CEEInfo::getFunctionFixedEntryPoint(CORINFO_METHOD_HANDLE   ftn,
+    // adheres to the limitations.
+    if (pMD->HasNativeCallableAttribute())
+    {
+        if (!pMD->IsStatic())
```
This validation should kick in for x86 as well

AaronRobinsonMSFT replied:
That is handled in the COM delegate call above.

AaronRobinsonMSFT committed:
Missing comma.

AaronRobinsonMSFT committed:
Remove duplicate entries in ReadyToRunHelper enum.

AaronRobinsonMSFT commented:
@janvorli Can you take a look at the work here and offer your thoughts on exceptions handling at the margins. @jkotas and I are in agreement that on Unix exceptions here aren't supported at this transition point, but on Windows we would want the typical behavior reverse P/Invoke behavior. Can you point me at an example of what I would need to do? Also, do you have any thoughts on the cost to support x86?

MichalStrehovsky commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -9202,18 +9202,40 @@ void CEEInfo::getFunctionFixedEntryPoint(CORINFO_METHOD_HANDLE   ftn,
+            COMPlusThrow(kNotSupportedException, W("NotSupported_NonStaticMethod"));
+
+        // No generic methods
+        if (pMD->IsGenericMethodDefinition())
```
Is this a check for uninstantiated generic method?

Shouldn't this be checking for instantiated generic methods and methods on generic types instead?

AaronRobinsonMSFT replied:
Yep. Was adding tests and discovered this as well. Thanks!

AaronRobinsonMSFT committed:
Validate instantiated generic methods for ldftn.

AaronRobinsonMSFT committed:
Add test cases for instantiated generic methods and call from unknown native thread.

MichalStrehovsky commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -9202,18 +9202,40 @@ void CEEInfo::getFunctionFixedEntryPoint(CORINFO_METHOD_HANDLE   ftn,
+            COMPlusThrow(kNotSupportedException, W("NotSupported_NonStaticMethod"));
+
+        // No generic methods
+        if (pMD->HasMethodInstantiation())
```
Should this rather be `HasClassOrMethodInstantiation`? Do we want to support this on generic types? At minimum `[NativeCallable(EntryPoint="Foo")]` would seem pretty difficult to support on generic types.

AaronRobinsonMSFT replied:
Yeah. Sorry about that. I just realized the issues with this one. Thanks.

AaronRobinsonMSFT committed:
Disallow non-generic methods from a generic class to be NativeCallable

AaronRobinsonMSFT committed:
Add test for generic class with NativeCallable static method.

AaronRobinsonMSFT committed:
Remove SAL annocations from NativeCallableAttribute test assets.

AaronRobinsonMSFT committed:
Move NativeCallableAttribute check to getCallInfo() in JIT interface.

AaronRobinsonMSFT committed:
Add test for non static methods for NativeCallableAttribute

AaronRobinsonMSFT committed:
[WIP] Check in CrossGen2 for NativeCallableAttribute usage.

janvorli commented:
@AaronRobinsonMSFT as for the EH on Linux, to ensure that we fail fast if there is an exception leaking from a method marked with NativeCallable, I can see two options here:
1. Modify UnwindManagedExceptionPass1 so that when it crosses the managed to native boundary, it checks if the managed method has native callable attribute and fail fast the process with unhandled exception in that case. The change would be trivial, but I am not sure about the cost of of the call to MethodDesc::HasNativeCallable at each managed to native boundary. @jkotas, do you think the cost of this call would be negligible here? 
2. Modify JIT so that it wraps bodies of methods marked with the NativeCallable attribute in a try / catch and calls FailFast in the catch.

AaronRobinsonMSFT commented:
Thanks @janvorli. (1) seems like a generally less ideal approach - doesn't really handle the pay-for-play mentality we try to have in the interop space. I had been thinking about (2) as an option but wanted to hear from you first. Since you suggested it, I will reach out to the @dotnet/jit-contrib team and ask for some advice if flags already exist or how I should think about the work. Would (2) apply for both Windows and non-Windows?

jkotas commented:
I think that (1) is the right way to do it. (2) adds performance overhead for no good reason.

> if the managed method has native callable attribute

This fact is encoded in GCInfo. So all you need to do is to crack GCInfo of the method that is not a big deal on exception handing path. .NET Native / CoreRT - that has very pay-for-play and performance minded design - does exactly that.

janvorli commented:
> This fact is encoded in GCInfo

Ah, I have forgotten that. It is the reverse pinvoke frame stack slot info presence, right?

jkotas commented:
> It is the reverse pinvoke frame stack slot info presence, right?

Yep

AaronRobinsonMSFT commented:
>I think that (1) is the right way to do it. (2) adds performance overhead to the non-exception path for no good reason.

@jkotas Doesn't (1) imply that whenever an exception is thrown this check can occur? As opposed to (2) that will only occur if a reverse P/Invoke happens. I would assume that managed exceptions occur far more often than a reverse P/Invoke or am I missing something here?

AaronRobinsonMSFT commented:
Update for PR. Spoke offline with @jkotas. The gist of the argument is that the exception path is already expensive and this check in the suggested area isn't going to move the needle in a measurable way; ergo (1) is appropriate.

janvorli commented:
@AaronRobinsonMSFT I've made the change in the UnwindManagedExceptionPass1 and I was really simple. I can either share the diff or I can make it a separate PR after your changes get in.

AaronRobinsonMSFT commented:
@janvorli W00t! Thank you much. If you can share the diff I will apply it.

janvorli committed:
Implement Unix EH change for NativeCallable methods

janvorli committed:
Implement Unix EH change for NativeCallable methods

AaronRobinsonMSFT committed:
Publicly expose NativeCallableAttribute.

AaronRobinsonMSFT committed:
Add testing for exceptions on Windows.

AaronRobinsonMSFT committed:
Add validation to CrossGen2.

AaronRobinsonMSFT committed:
Update test.

AaronRobinsonMSFT commented:
@jkotas @janvorli @MichalStrehovsky @jkoritzinsky Please take another look. I believe I have addressed all outstanding issues and added sufficient scenario testing.

jkotas commented:
This is a performance optimization. Do you have any performance numbers?

jkotas commented on src/coreclr/src/vm/jithelpers.cpp:
```diff
@@ -5133,6 +5133,33 @@ Thread * __stdcall JIT_InitPInvokeFrame(InlinedCallFrame *pFrame, PTR_VOID StubS
+
+    Thread* thread = GetThreadNULLOk();
+    if (thread == NULL)
+        thread = CreateThreadBlockThrow();
```
There is a probably a way to make this faster by moving all rare cases into a separate `JIT_ReversePInvokeEnterRare` method.

AaronRobinsonMSFT replied:
Yes, you were right. The amount of generated code was substantially reduced by reorganizing the branches.

jkotas commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -9202,18 +9218,25 @@ void CEEInfo::getFunctionFixedEntryPoint(CORINFO_METHOD_HANDLE   ftn,
+    // time permits investigation into all the potential issues.
     if (pMD->HasNativeCallableAttribute())
     {
         pResult->addr = (void*)COMDelegate::ConvertToCallback(pMD);
```
Can we make `COMDelegate::ConvertToCallback` ifdefed out with the same condition?

AaronRobinsonMSFT replied:
As in removing the function entirely? The linker would normally remove a function if it isn't used.

jkotas replied:
It is more just for clarify that the function is used for Windows x86 only.

jkotas commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -5095,6 +5095,22 @@ void CEEInfo::getCallInfo(
+    // then validate it adheres to the limitations.
+    if ((flags & CORINFO_CALLINFO_LDFTN) && pMD->HasNativeCallableAttribute())
+    {
+        if (!pMD->IsStatic())
```
Can we delete this check from ``COMDelegate::ConvertToCallback` ?

AaronRobinsonMSFT replied:
The `COMDelegate::ConvertToCallback` is called in a scenario below. However, I think that should be updated as well right?

https://github.com/dotnet/runtime/blob/ecd8378926fa27c5ee838ff0c5e67798f88e127e/src/coreclr/src/vm/jitinterface.cpp#L13353-L13361

jkotas replied:
Right

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeCallableAttribute.cs:
```diff
@@ -5,10 +5,16 @@
     /// <summary>
-    /// Any method marked with NativeCallableAttribute can be directly called from
+    /// Any method marked with <see cref="System.Runtime.InteropServices.NativeCallableAttribute" /> can be directly called from
     /// native code. The function token can be loaded to a local variable using LDFTN
```
> The function token can be loaded to a local variable using LDFTN
 
This is not a good user-facing documentation. We should write this snippet as if the C# function pointers feature existed already.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeCallableAttribute.cs:
```diff
@@ -17,12 +23,12 @@ public NativeCallableAttribute()
 
         /// <summary>
-        /// Optional. If omitted, then the method is native callable, but no EAT is emitted.
+        /// Optional. If omitted, then the method is native callable, but no export is emitted during AOT compilation.
```
```suggestion
        /// Optional. If omitted, then the method is native callable, but no export is emitted during compilation.
```
This is not specific to AOT

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeCallableAttribute.cs:
```diff
@@ -17,12 +23,12 @@ public NativeCallableAttribute()
 
         /// <summary>
-        /// Optional. If omitted, compiler will choose one for you.
+        /// Optional. If omitted, the runtime will choose one for you based on the platform.
```
```suggestion
        /// Optional. If omitted, the runtime will use the default platform calling convention.
```

AaronRobinsonMSFT committed:
Update src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeCallableAttribute.cs

Co-Authored-By: Jan Kotas <jkotas@microsoft.com>

AaronRobinsonMSFT committed:
Update src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeCallableAttribute.cs

Co-Authored-By: Jan Kotas <jkotas@microsoft.com>

jkotas commented on src/coreclr/src/vm/exceptionhandling.cpp:
```diff
@@ -4664,8 +4663,26 @@ VOID DECLSPEC_NORETURN UnwindManagedExceptionPass1(PAL_SEHException& ex, CONTEXT
             controlPc = Thread::VirtualUnwindLeafCallFrame(frameContext);
         }
 
+        bool isManagedCode = ExecutionManager::IsManagedCode(controlPc);
```
This condition does not look right to me. Why do we check `!isManagedCode` here?

janvorli replied:
This detects the boundary when we've unwound from a managed to a native frame. If the NativeCallable method was the managed one, the native frame is its caller and that's when we need to fail. The codeInfo still has info on the managed frame.

jkotas replied:
What happens in the case when the caller is managed code (unmanaged calli)? I think we should also fail with unhandled exception in this case.

jkotas commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -10091,7 +10114,7 @@ void CEEInfo::getEEInfo(CORINFO_EE_INFO *pEEInfoOut)
     pEEInfoOut->offsetOfWrapperDelegateIndirectCell = OFFSETOF__DelegateObject__methodPtrAux;
 
-    pEEInfoOut->sizeOfReversePInvokeFrame  = (DWORD)-1;
+    pEEInfoOut->sizeOfReversePInvokeFrame = sizeof(FastReversePInvokeFrame);
```
For crossgen, we should make this a constant `READYTORUN_ReversePInvokeTransitionFrameSizeInPointerUnits`, and follow the same pattern for it as for `READYTORUN_PInvokeTransitionFrameSizeInPointerUnits`.

This constant should include some extra space to give us version resiliency flexibility, same as `READYTORUN_PInvokeTransitionFrameSizeInPointerUnits`. I think making `2` would be reasonable.

MichalStrehovsky commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -5095,6 +5095,22 @@ void CEEInfo::getCallInfo(
+    if ((flags & CORINFO_CALLINFO_LDFTN) && pMD->HasNativeCallableAttribute())
+    {
+        if (!pMD->IsStatic())
+            EX_THROW(EEResourceException, (kNotSupportedException, W("NotSupported_NonStaticMethod")));
```
Documentation for NotSupportedException says: "The exception that is thrown when an invoked method is not supported"

Documentation for InvalidProgramException says: "The exception that is thrown when a program contains invalid Microsoft intermediate language (MSIL) or metadata."

This is invalid metadata - feels like InvalidProgram is more appropriate.

```suggestion
            EX_THROW(EEResourceException, (kInvalidProgramException, W("NotSupported_NonStaticMethod")));
```

AaronRobinsonMSFT replied:
@MichalStrehovsky I was thinking about this after our offline conversation. I don't think the metadata is wrong. The current limitations are largely an implementation detail and there is nothing wrong with the metadata since it is valid/parsable. I would argue that the `InvalidProgramException` should not be thrown if the compiler emits it. This isn't to say the compiler can't emit an invalid program, but then it would be a compiler bug.

MichalStrehovsky replied:
Right - we'll want the compilers to be aware of `NativeCallable` methods so that they disallow this and other problematic patterns (e.g. a managed call to a `NativeCallable` method).

To me this feels like it's in the same bucket as e.g. `DllImport` - compilers validate that it's not used illegally and if it is, it's a compiler bug.

Placing DllImport on an instance method will also produce an `InvalidProgramException` (if you bypass validating compilers and go directly to IL):

```
Unhandled Exception: System.InvalidProgramException: Error: PInvoke item (field,method) must be Static.
   at Blah.MessageBox(IntPtr a, IntPtr b, IntPtr c, IntPtr d)
   at Program.Main()
```

It's a behavior that we didn't define and the metadata is therefore not just unsupported - it's invalid.

AaronRobinsonMSFT replied:
>we'll want the compilers to be aware of NativeCallable methods so that they disallow this and other problematic patterns (e.g. a managed call to a NativeCallable method).

Yep. I reached out to @jaredpar about having Roslyn do some work - https://github.com/dotnet/runtime/issues/32462#issuecomment-588508231. I think you have again convinced me that this should be `InvalidProgramException`. Thanks for entertaining my thoughts.

MichalStrehovsky commented on src/coreclr/src/tools/Common/TypeSystem/Common/ExceptionStringID.cs:
```diff
@@ -34,6 +34,7 @@ public enum ExceptionStringID
         InvalidProgramNativeCallable,
         InvalidProgramCallAbstractMethod,
         InvalidProgramCallVirtStatic,
+        InvalidProgramNativeCallableAttributeUsage,
```
Nit: Could you add IDs for all 3 variants of these messages? While warnings from crossgen are not super important (but we will want to surface them), the type system is eventually going to be used by the ILVerify tool once it's moved over (#13827) and it shouldn't be that we get more descriptive error message from the runtime than from the tool.

MichalStrehovsky commented on src/coreclr/src/tools/Common/TypeSystem/Common/ThrowHelper.cs:
```diff
@@ -42,6 +42,12 @@ public static void ThrowInvalidProgramException()
         }
 
+        [System.Diagnostics.DebuggerHidden]
+        public static void ThrowInvalidProgramException(ExceptionStringID id)
```
Nit: the existing overload that also includes the method with the problem seems to be better, diagnostic-wise.

MichalStrehovsky commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -1145,6 +1151,17 @@ private void getFieldInfo(ref CORINFO_RESOLVED_TOKEN pResolvedToken, CORINFO_MET
+            {
+                if (!originalMethod.Signature.IsStatic // Must be a static method
+                    || originalMethod.HasInstantiation || originalMethod.OwningType.HasInstantiation // No generics involved
+                    || Marshaller.IsMarshallingRequired(originalMethod)) // Only blittable arguments
```
I think this needs to be the `Marshaller.IsMarshallingRequired(originalMethod.Signature)` overload. The one that takes `MethodDesc` expects the MethodDesc to be a p/invoke and asserts that as the first thing.

AaronRobinsonMSFT committed:
Error cases for NativeCallableAttribute should be InvalidProgramException
  instead of NotSupportedException.

AaronRobinsonMSFT committed:
Create ReadyToRun constant for Reverse PInvoke frame.
COMDelegate::ConvertToCallback is not only compiled on Win-x86 builds.

AaronRobinsonMSFT committed:
Documentation update related to future C# language feature.

AaronRobinsonMSFT committed:
Revert "Implement Unix EH change for NativeCallable methods"

This reverts commit 2dbb74803f890ba5fbaf8158f1d98b5aa6776246.

AaronRobinsonMSFT committed:
Clean up tests

AaronRobinsonMSFT committed:
Create a JIT_ReversePInvokeEnterRare() implementation for the slower path
  during a reverse PInvoke enter.

jkotas commented on src/coreclr/src/vm/jithelpers.cpp:
```diff
@@ -5028,6 +5028,56 @@ Thread * __stdcall JIT_InitPInvokeFrame(InlinedCallFrame *pFrame, PTR_VOID StubS
+EXTERN_C void STDCALL ReversePInvokeBadTransition();
+
+// This is a slower version of the reverse PInvoke enter function.
+static void JIT_ReversePInvokeEnterRare(FastReversePInvokeFrame* frame)
```
We typically mark rare paths like this with `NOINLINE` to make it clear to the C++ compiler that we really want to keep the hot and rare paths separate.

AaronRobinsonMSFT committed:
Move exception testing to Windows only since non-Windows will failfast.

AaronRobinsonMSFT committed:
Mark rare path as NOINLINE.

AaronRobinsonMSFT commented:
>This is a performance optimization. Do you have any performance numbers?

@jkotas We have a slight win here. Looking in a profiler, the stack is shallower and more direct. I am also including a screenshot of the sample breakdown for the `JIT_ReversePInvokeEnter` function. 

Iteration count = 100000000
```
GetFunctionPointerForDelegate: - 15.995048 ns/iter
NativeCallableAttribute: - 14.440174 ns/iter
GetFunctionPointerForDelegate: - 15.770667 ns/iter
NativeCallableAttribute: - 14.415601 ns/iter
```

![image](https://user-images.githubusercontent.com/30635565/76150228-ec8ad280-605c-11ea-89ad-fb425ae19691.png)

AaronRobinsonMSFT commented:
Seems like we have an issue in TieredCompilation in the call count logic. Since we are doing a reverse P/Invoke it is possible the thread doesn't exist. We can do one of two things here.

1) Skip call counting if a thread doesn't exist.
1) Update the below logic to create a thread if it doesn't exist.

I think that (2) is probably the right thing to do since the call is going to have to create a thread anyways. Opinions?

https://github.com/dotnet/runtime/blob/5c981af975915b87c68577bbb9be098bfee203c0/src/coreclr/src/vm/callcounting.cpp#L704-L729

AaronRobinsonMSFT commented:
@kouvel do you have any opinion on this issue - https://github.com/dotnet/runtime/pull/33005#issuecomment-596125489.

AaronRobinsonMSFT committed:
Add support to CallCountingManager for Preemptive threads.

jkotas commented on src/coreclr/src/vm/callcounting.cpp:
```diff
@@ -777,6 +788,30 @@ PCODE CallCountingManager::OnCallCountThresholdReached(TransitionBlock *transiti
+    // that the thread is new to the runtime so we might have to
+    // create one.
+    if (currentThreadMaybe == NULL)
+        currentThreadMaybe = CreateThreadBlockThrow();
```
If this throws on Unix, where is this exception going to be caught?

AaronRobinsonMSFT replied:
Boo. It won't be. Is this one of those scenarios where I should attempt to create and if I can't fail fast?

jkotas replied:
Right. Make sure that it is diagnosable from what we print to stdout. Ideally, it would be same as any other unhanded managed exception.

AaronRobinsonMSFT replied:
Wouldn't this also mean that the same call in `PreStubWorker_Preemptive()` needs to do something different? I don't think that is the case because the `ThePreStub()` itself should have already installed a frame right - I want to make sure I am understanding the difference in these code paths.

jkotas replied:
Try it and see what it does...

There macros in `ThePreStub()` are doing something, but I am not sure whether it is the right thing (on both Windows and Unix):
```
INSTALL_MANAGED_EXCEPTION_DISPATCHER;
INSTALL_UNWIND_AND_CONTINUE_HANDLER;
```

AaronRobinsonMSFT replied:
Updated the call in `PreStubWorker_Preemptive()` because that call was not protected. Unfortunately moving the following:

```
INSTALL_MANAGED_EXCEPTION_DISPATCHER;
INSTALL_UNWIND_AND_CONTINUE_HANDLER;
```

wasn't going to work because those macros depend on `CURRENT_THREAD`. Therefore I am calling the no throw option and will call `TerminateProcess()` if the thread is still `NULL`.

jkotas commented:
> @jkotas We have a slight win here. 

That is less than I would expect. We should be able to do better. The dissasembly that I see on Win x64 is:

```
CoreCLR!JIT_ReversePInvokeEnter:
00007fff`8d9084f0 4053            push    rbx
00007fff`8d9084f2 4883ec20        sub     rsp,20h
00007fff`8d9084f6 488bd9          mov     rbx,rcx
00007fff`8d9084f9 e8d225faff      call    CoreCLR!GetThread (00007fff`8d8aaad0)
00007fff`8d9084fe 4885c0          test    rax,rax
00007fff`8d908501 7421            je      CoreCLR!JIT_ReversePInvokeEnter+0x34 (00007fff`8d908524)
00007fff`8d908503 8b500c          mov     edx,dword ptr [rax+0Ch]
00007fff`8d908506 85d2            test    edx,edx
00007fff`8d908508 751a            jne     CoreCLR!JIT_ReversePInvokeEnter+0x34 (00007fff`8d908524)
00007fff`8d90850a c7400c01000000  mov     dword ptr [rax+0Ch],1
00007fff`8d908511 8b15791e3500    mov     edx,dword ptr [CoreCLR!g_TrapReturningThreads (00007fff`8dc5a390)]
00007fff`8d908517 85d2            test    edx,edx
00007fff`8d908519 7509            jne     CoreCLR!JIT_ReversePInvokeEnter+0x34 (00007fff`8d908524)
00007fff`8d90851b 488903          mov     qword ptr [rbx],rax
00007fff`8d90851e 4883c420        add     rsp,20h
00007fff`8d908522 5b              pop     rbx
00007fff`8d908523 c3              ret
00007fff`8d908524 488bcb          mov     rcx,rbx
00007fff`8d908527 4883c420        add     rsp,20h
00007fff`8d90852b 5b              pop     rbx
00007fff`8d90852c e967ffffff      jmp     CoreCLR!JIT_ReversePInvokeEnterRare (00007fff`8d908498)
```

We are losing cycles on non-inlined `CoreCLR!GetThread`. Can you look into why it is not getting inlined? Are we missing `#include "threads.inl" somewhere?

AaronRobinsonMSFT commented:
>We are losing cycles on non-inlined CoreCLR!GetThread. Can you look into why it is not getting inlined? Are we missing `#include "threads.inl" somewhere?

I saw the same disassembly. I didn't realize `GetThread()` should be inlined. Stepping through with the debugger, we are stepping into `thread.inl`. I will see about other cases.

AaronRobinsonMSFT commented:
Unclear why that function isn't being inlined, but the results are largely the same with the following disassembly. I manually inlined the contents of `GetThread()`.

```
00007FF9E6797210 8B 15 E2 58 34 00    mov         edx,dword ptr [_tls_index (07FF9E6ADCAF8h)]  
00007FF9E6797216 65 48 8B 04 25 58 00 00 00 mov         rax,qword ptr gs:[58h]  
00007FF9E679721F 41 B8 08 00 00 00    mov         r8d,8  
00007FF9E6797225 48 8B 04 D0          mov         rax,qword ptr [rax+rdx*8]  
00007FF9E6797229 49 8B 14 00          mov         rdx,qword ptr [r8+rax]  
00007FF9E679722D 48 85 D2             test        rdx,rdx  
00007FF9E6797230 74 1C                je          JIT_ReversePInvokeEnter+3Eh (07FF9E679724Eh)  
00007FF9E6797232 8B 42 0C             mov         eax,dword ptr [rdx+0Ch]  
00007FF9E6797235 85 C0                test        eax,eax  
00007FF9E6797237 75 15                jne         JIT_ReversePInvokeEnter+3Eh (07FF9E679724Eh)  
00007FF9E6797239 C7 42 0C 01 00 00 00 mov         dword ptr [rdx+0Ch],1  
00007FF9E6797240 8B 05 CA E1 34 00    mov         eax,dword ptr [g_TrapReturningThreads (07FF9E6AE5410h)]  
00007FF9E6797246 85 C0                test        eax,eax  
00007FF9E6797248 75 04                jne         JIT_ReversePInvokeEnter+3Eh (07FF9E679724Eh)  
00007FF9E679724A 48 89 11             mov         qword ptr [rcx],rdx  
00007FF9E679724D C3                   ret  
00007FF9E679724E E9 61 FF FF FF       jmp         JIT_ReversePInvokeEnterRare (07FF9E67971B4h)
```

kouvel commented:
> Seems like we have an issue in TieredCompilation in the call count logic. Since we are doing a reverse P/Invoke it is possible the thread doesn't exist.

Is this an issue that would exist before this change or only after this change?

My first thought is that the reverse-pinvoke transition would be the best place to create the thread object. Would that be possible? I'm not aware if that is something that is being tried to be avoided.

If this is expected after this change, I'm curious about other potential issues with the thread object not existing.

> I think that (2) is probably the right thing to do since the call is going to have to create a thread anyways.

Where would the thread object be created?

If it's reaching this path without a thread object available, it can also reach jitted code of the method without the thread object being available. I'm not sure if there would be other issues with that. Just an example, how would the following not be possible:
- The native-callable method is already jitted (on a different thread)
- The native code calls the method from the new thread after it has already been jitted, which would typically go to the jitted code without going through the prestub
- The managed native-callable method calls a virtual method that ends up in `VSD_ResolveWorker`, which may have similar issues with not having the thread object available - not entirely sure

I don't know if it is currently safe to call managed code on a thread without a thread object, perhaps others would know. From briefly looking at the relevant code, it almost seems like a thread object is necessary to transition between preemptive and cooperative GC modes, though that's just a far estimation I could be wrong about that.

kouvel commented:
In other words shouldn't the thread object already be created before reaching this code?

AaronRobinsonMSFT commented:
Thanks @kouvel. This is entirely related to my current change and is new behavior. Please take a look at how I fixed the issue. @jkotas called out an issue with [exceptions](https://github.com/dotnet/runtime/pull/33005#discussion_r389306826) that I will address, but does my proposed approach jive with your style existing design okay?

https://github.com/dotnet/runtime/blob/6e10c691cf537913d768d2b8f7ffebf3c0edb11f/src/coreclr/src/vm/callcounting.cpp#L698-L712

kouvel commented:
A potential issue is that in the call sequence that would lead to the `OnCallCountThresholdReached` stub - precode => call counting stub => `OnCallCountThresholdReached`, the call counting stub pointer may be invalidated if that call sequence doesn't happen in cooperative GC mode. Call counting stubs are deleted and the current expectation is that the parameter, `stubIdentityingToken`, which is an address into the call counting stub, can be safely used in `OnCallCountThresholdReached` to read some information because it would be in cooperative GC mode, during which the call counting stub pointer would be guaranteed to be valid. So as it is currently, the cooperative mode transition would have to occur before the method's entry point is called.

jkotas commented:
Is there a way to pin the call counting stubs? The easiest option may be to disable tiered JIT for these and always JIT them as tier 1.

kouvel commented:
> Is there a way to pin the call counting stubs?

Not at the moment, but I don't think it would be difficult to add if necessary.

Aside from tiered compilation, would it be ok to call jitted code of a managed method without a cooperative mode transition?

kouvel commented:
Pinning call counting stubs would probably involve adding another layer of locking of some sort to synchronize with deletion, not much different from the cooperative GC mode transition

AaronRobinsonMSFT commented:
>Aside from tiered compilation, would it be ok to call jitted code of a managed method without a cooperative mode transition?

No. We are handling that transition in a JIT helper:

https://github.com/dotnet/runtime/blob/6e10c691cf537913d768d2b8f7ffebf3c0edb11f/src/coreclr/src/vm/jithelpers.cpp#L5051-L5071

kouvel commented:
Ah ok, I think I understand now how it's intended to work. I agree it would be easiest to disable tiered compilation for native-callable methods.

kouvel commented:
If there is a cheap check that can be done it can be disabled in `MethodDesc::DetermineAndSetIsEligibleForTieredCompilation` (this method is called for a large number of methods during startup).

AaronRobinsonMSFT commented:
Thanks @kouvel. Will try that out.

AaronRobinsonMSFT committed:
Revert "Add support to CallCountingManager for Preemptive threads."

This reverts commit 6e10c691cf537913d768d2b8f7ffebf3c0edb11f.

AaronRobinsonMSFT committed:
Disable NativeCallableAttribute for TieredCompilation consideration.

AaronRobinsonMSFT committed:
Handle the failure to create Thread in the reverse P/Invoke case.

jkotas commented on src/coreclr/src/vm/prestub.cpp:
```diff
@@ -1766,112 +1766,173 @@ extern "C" MethodDesc * STDCALL PreStubGetMethodDescForCompactEntryPoint (PCODE
+        // going to work.
+        if (currentThread == NULL)
+        {
+            TerminateProcess(GetCurrentProcess(), hr);
```
How are people going to figure what went wrong when they hit this?

andrew-boyarshin replied:
I feel like this could use a proper failfast, since that's what it was designed for. But yeah, logging wouldn't hurt too.

AaronRobinsonMSFT replied:
I don't know the best way to author a failfast. I was copying places in the exception code paths i know are failfast. If there is a common pattern could someone point it out?

andrew-boyarshin replied:
`EEPOLICY_HANDLE_FATAL_ERROR`? That's what managed code ends up in. And it is used in vm code too.

AaronRobinsonMSFT replied:
Thanks for the pointer. Looks like `EEPOLICY_HANDLE_FATAL_ERROR_WITH_MESSAGE()` is the most appropriate option here.

jkotas commented on src/coreclr/src/vm/method.cpp:
```diff
@@ -4839,6 +4839,10 @@ bool MethodDesc::DetermineAndSetIsEligibleForTieredCompilation()
 
+        // Functional requirement of being in COOP mode during a call count query.
+        // NativeCallable methods are, by definition, not initiated from COOP mode.
+        !HasNativeCallableAttribute() &&
```
Is there a way to do this lazily? This going to make us lookup this NativeCallableAttribute for a lot.

jkotas replied:
(Custom attribute lookup is not a cheap operation.)

AaronRobinsonMSFT replied:
Agreed on cost. I ran it under a profiler and didn't observe any hits, but admit that doesn't mean it isn't expensive. Is this something that would better as a flag that we lookup and set once? When you say lazy do you mean closer to decision time? I thought this was that point.

AaronRobinsonMSFT replied:
Well the issue I am seeing is one of truth. The `MethodDesc::IsEligibleForTieredCompilation()` call is made in many places, which means I could only update those locations that would respond to this attribute - I assume this is what is meant by lazy. However then other callers of this would be lied to since the flag in question was set. This leads me to believe that what we should do is look it up once and set the flag and never have to look it up again. Does that align with your concerns here?

jkotas replied:
We are able to flags method as non-eligible for tiered compilation lazily when the method is JITed / executed for the first time (e.g. look for `JitSwitchedToOptimized()`). I think this should do something similar.

AaronRobinsonMSFT replied:
Wow. There are lots of options here. I could also tell the JIT when it sees `JIT_FLAG_REVERSE_PINVOKE` to call back in and tell the runtime to reset the method to `CORINFO_FLG_SWITCHED_TO_OPTIMIZED`. This would move the query for `NativeCallableAttribute` to the absolute latest time. I will investigate a few avenues. Thanks!

jkotas replied:
> tell the JIT when it sees JIT_FLAG_REVERSE_PINVOKE to call back in and tell the runtime to reset the method to CORINFO_FLG_SWITCHED_TO_OPTIMIZED

We try to avoid having runtime-specific policies like this hard-coded in the JIT.

AaronRobinsonMSFT replied:
Okay. I will keep the policy side of things in the runtime. I have ideas that definitely avoid changes in the JIT.

AaronRobinsonMSFT replied:
Okay. I will keep the policy side of things in the runtime. I have ideas that definitely avoid changes in the JIT.

kouvel replied:
One way might be to pass `PrepareCodeConfig *` into `UnsafeJitFunction` instead of the `NativeCodeVersion` (which can be obtained from `PrepareCodeConfig`), and after the call to `GetCompileFlags()`, if `CORJIT_FLAG_REVERSE_PINVOKE` is set, remove the flag `CORJIT_FLAG_TIER0` and call `pConfig->SetJitSwitchedToOptimized()`, which would cause tiering to be disabled for the method in `JitCompileCodeLocked`.

jkotas replied:
Yes, something like this should work. Also, we need to have an equivalent for R2R (this path won't kick in for R2R).

kouvel replied:
Oh right. That could be done in `MethodDesc::GetPrecompiledCode`, something like this might work (the tier has to be changed before `SetNativeCode` to avoid races):
```c++
        pCode = GetPrecompiledR2RCode(pConfig);
        if (pCode != NULL)
        {
            LOG_USING_R2R_CODE(this);

        #ifdef FEATURE_TIERED_COMPILATION
            bool shouldTier = pConfig->GetMethodDesc()->IsEligibleForTieredCompilation();
        #if !defined(TARGET_X86) || !defined(TARGET_WINDOWS)
            if (shouldTier && HasNativeCallableAttribute())
            {
                // These methods are not tiered, update the tier in the code version
                NativeCodeVersion codeVersion = pConfig->GetCodeVersion();
                if (codeVersion.IsDefaultVersion())
                {
                    pConfig->GetMethodDesc()->GetLoaderAllocator()->GetCallCountingManager()->DisableCallCounting(codeVersion);
                }
                codeVersion.SetOptimizationTier(NativeCodeVersion::OptimizationTierOptimized);
                shouldTier = false;
            }
        #endif // !TARGET_X86 || !TARGET_WINDOWS
        #endif // FEATURE_TIERED_COMPILATION

            if (pConfig->SetNativeCode(pCode, &pCode))
            {
            #ifdef FEATURE_CODE_VERSIONING
                pConfig->SetGeneratedOrLoadedNewCode();
            #endif
            #ifdef FEATURE_TIERED_COMPILATION
                if (shouldTier)
                {
                    _ASSERTE(pConfig->GetCodeVersion().GetOptimizationTier() == NativeCodeVersion::OptimizationTier0);
                    pConfig->SetShouldCountCalls();
                }
            #endif
            }
        }
```

jkotas replied:
It would be nice to avoid custom attribute lookups on the R2R path and encode this bit in the R2R fixup instead.

AaronRobinsonMSFT replied:
@jkotas and @kouvel I have been able to apply the suggestion to `UnsafeJitFunction()`. I will need some guidance on the R2R side of things. I have found the `enum ReadyToRunFixupKind` in `readytorun.h`. Would the approach be to insert a bool into the readytorun metadata to indicate the `NativeCallableAttribute`?

An existing example or perhaps a pointer to some guide would be helpful here. My biggest issue is understanding how to query the R2R embedded data and exactly what data is interesting to embed.

kouvel replied:
I'm not very familiar with how that should be implemented, @MichalStrehovsky would you be able to help?

MichalStrehovsky replied:
> I'm not very familiar with how that should be implemented, @MichalStrehovsky would you be able to help?

I haven't really done anything with R2R fixups and didn't spend much time reviewing that code either. Hopefully @trylek will be able to point you in the right direction. If not, I'll spend some time reading that code and come up with pointers.

AaronRobinsonMSFT committed:
Use EEPOLICY_HANDLE_FATAL_ERROR_WITH_MESSAGE() for failfast.

AaronRobinsonMSFT committed:
Revert "Disable NativeCallableAttribute for TieredCompilation consideration."

This reverts commit c0848abffcfe79cde78ed3402da03703c35c4195.

AaronRobinsonMSFT committed:
Disable Tiered Compilation for NativeCallableAttribute methods.

AaronRobinsonMSFT commented:
@MichalStrehovsky See [here](https://github.com/dotnet/runtime/pull/33005#discussion_r390119546) for all the details, but I think I see how this works now.

My guess is I should add a new value to `CORCOMPILE_FIXUP_BLOB_KIND` and `ReadyToRunFixupKind` - also update crossgen2 - and store whatever data I want in that blob and it can be read out in `DynamicHelperWorker()` in `prestub.cpp`. From here I can store that data ... somewhere? That is the part I am having trouble understanding - can I query this information whenever I want?

jkotas commented on src/coreclr/src/vm/jithelpers.cpp:
```diff
@@ -5028,6 +5028,56 @@ Thread * __stdcall JIT_InitPInvokeFrame(InlinedCallFrame *pFrame, PTR_VOID StubS
+    _ASSERTE(frame != NULL);
+    _ASSERTE(frame->currentThread == GetThread());
+
+    frame->currentThread->EnablePreemptiveGC();
```
It should be sufficient to just clear `m_fPreemptiveGCDisabled` like what the current delegate assembly helpers are doing. The extra `CatchAtSafePoint` should not be required.

AaronRobinsonMSFT replied:
@jkotas I get the reason for the `*Enter()` case being unwrapped because of the fallback (i.e. `*Rare()`) that will handle this but the `*Exit()` doesn't have that. What is the reason this isn't needed? Personally, accessing fields in this way when public functions exist is a code smell and calls into question why these functions exist if we don't use them consistently. Perhaps this warrants a new function `EnablePreemptiveGCFast()`? I accept that your statement is correct to just set the bit, but why?

jkotas replied:
> What is the reason this isn't needed? 

For the unmanaged/managed transitions, the effect of CatchAtSafePoint would be that the GC suspension will finish a bit faster in some cases. It is a trade-off 

It is a trade-off: Spend extra cycles on each transition and make GC suspension faster; or make each transition faster and make GC suspension a bit slower occasionally.

We are opting for the faster unmanaged/managed transitions everywhere else today, so this is for consistency. Also, the existing unmanaged/managed transitions are in assembly helpers, so there was no need to have a C++ method to do that.

> EnablePreemptiveGCFast

I do not think that having `EnablePreemptiveGCFast` method would be an improvement over just setting the field directly.

If you would like to make this pretty, the way to do that would be to encapsulate the transition in FORCEINLINE methods, similar to how it is done in CoreRT: https://github.com/dotnet/corert/blob/41557532f0ff3e4bb0393a5451cae65a041ad5ae/src/Native/Runtime/thread.h#L259

AaronRobinsonMSFT committed:
Change name of new reverse P/Invoke frame to match CoreRT.

AaronRobinsonMSFT committed:
Manually inline enable preemptive.

AaronRobinsonMSFT commented:
Update performance numbers are below. Note that this include the [fix for the MSVC inline issue](https://github.com/dotnet/runtime/pull/33347).

Iteration count = 100000000

```
GetFunctionPointerForDelegate: - 14.905451 ns/iter
NativeCallableAttribute: - 13.273389 ns/iter
GetFunctionPointerForDelegate: - 14.874898 ns/iter
NativeCallableAttribute: - 13.620476 ns/iter
```

CrossGen'd image:
```
GetFunctionPointerForDelegate: - 14.933282 ns/iter
NativeCallableAttribute: - 13.038117 ns/iter
GetFunctionPointerForDelegate: - 14.829792999999999 ns/iter
NativeCallableAttribute: - 12.976431999999999 ns/iter
```

AaronRobinsonMSFT committed:
Handle issues for Tiered Compilation with NativeCallable in ReadyToRun images.

kouvel commented:
Commit https://github.com/dotnet/runtime/pull/33005/commits/c78f4a4288aeee32938e0f96535e52d0978bfb69 looks good to me

jkotas commented on src/coreclr/src/vm/codeversion.cpp:
```diff
@@ -1670,6 +1671,14 @@ PCODE CodeVersionManager::PublishVersionableCodeIfNecessary(
+
+#ifdef FEATURE_TIERED_COMPILATION
+            // If the caller is in preemptive mode, then
+            // Tiered Compilation isn't supported. (e.g. NativeCallable)
```
I believe that we can come here with `CallerGCMode_Unknown`, e.g. if somebody calls `RuntimeHelpers.PrepareMethod` on the method. Maybe add test for it?

If we come here with unknown, we need to lookup the attribute the expensive way.

AaronRobinsonMSFT replied:
Oh wow. I just realized what this means. Okay this definitely warrants a test. Thanks.

jkotas commented on src/coreclr/src/vm/codeversion.cpp:
```diff
@@ -1670,6 +1671,14 @@ PCODE CodeVersionManager::PublishVersionableCodeIfNecessary(
+            // If the caller is in preemptive mode, then
+            // Tiered Compilation isn't supported. (e.g. NativeCallable)
+            if (callerGCMode == MethodDesc::CallerGCMode_Preempt)
+                config->SetJitSwitchedToOptimized();
```
There are other places in the system that call PrepareCode. Does this need to be pushed into `PrepareCode`?

AaronRobinsonMSFT replied:
As in the propagation of the caller's GC mode?

kouvel replied:
There appears to be at least one relevant call here (error path), but there may be others in the future:
https://github.com/dotnet/runtime/commit/c78f4a4288aeee32938e0f96535e52d0978bfb69#diff-b06261f7127af3268df1f78173d20b34R1659

Wonder if it would be safer to put the enum value into `PrepareCodeConfig` and check for that after looking up R2R'ed code instead of using `JitSwitchedToOptimized`. If that value is Unknown (default value when not explicitly set, this would hopefully cover all other paths) then it can do the expensive lookup.

kouvel replied:
And that value could then be reused on the JIT path as well to avoid the expensive lookup there

AaronRobinsonMSFT replied:
So enrich `PrepareCodeConfig`? That might make the enum build issue a bit easier to address. Let me try that.

AaronRobinsonMSFT replied:
Well that isn't going to address the build issues. This is an absolute mess.

jkotas replied:
> that value could then be reused on the JIT path

I think we should always do the attribute lookup on the JIT path to get the right code generated. If we do not do that, I think the detection of incorrect use is not going to work well.

jkotas commented on src/coreclr/src/vm/method.hpp:
```diff
@@ -1788,7 +1786,14 @@ class MethodDesc
+        CallerGCMode_Coop,
+        CallerGCMode_Preempt    // (e.g. NativeCallableAttribute)
+    };
+    PCODE DoPrestub(MethodTable *pDispatchingMT, DWORD callerGCMode = CallerGCMode_Unknown);
```
Can this use actual enum name instead of DWORD?

AaronRobinsonMSFT replied:
I tried for about 20 minutes. It wasn't obvious how to do it if the enum was nested in `MethodDesc` - the type is forward declared all over and the header includes are a giant mess. I could pull out the enum and try putting it in something like `common.h`, but the issue there is naming and there is some annoyance that I didn't look into between `codeman.h` and `codeversion.h`. I can give it another try or if you have a suggestion on where where to attempt that would be appreciated.

AaronRobinsonMSFT replied:
This entire problem is related to `CodeVersionManager::PublishVersionableCodeIfNecessary()` and the header file declaring that function being included in some very targeted code paths. I can start throwing headers at it to make it build, but that I think that is how we got into the current mess. I need some distance from this. Will pick it up tomorrow.

AaronRobinsonMSFT committed:
Add test for code path involving RuntimeHelpers.PrepareMethod().

AaronRobinsonMSFT committed:
Handle additional locations for R2R scenarios.

kouvel commented:
Otherwise https://github.com/dotnet/runtime/pull/33005/commits/105d53f2d948b829f24eae4bd18acc16b54e50d5 looks good to me

AaronRobinsonMSFT commented:
@kouvel The diff link doesn't seem to point at an error case or I am just unsure how to interpret github navigation. Can you specifically point at the location in question?

@jkotas @MichalStrehovsky @janvorli  Any other feedback? I was going to add a final assert but wanted to wait if there was anything else first.

jkotas commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -13351,6 +13385,7 @@ BOOL LoadDynamicInfoEntry(Module *currentModule,
 
         MethodEntry:
+#if defined(TARGET_X86) && defined(TARGET_WINDOWS)
             if (kind == ENCODE_METHOD_NATIVE_ENTRY)
```
`ENCODE_METHOD_NATIVE_ENTRY` handling does look quite right:
- It is left over from fragile crossgen, not wired into R2R correctly (e.g. it is missing in inc/readytorun.h)
- It is not handled in R2RDump, crossgen2, etc.

I think it may be easiest to fail crossgen for anything related to native callable on Windows x86 for now (both AOTing the method itself, and also AOTing methods that load the address of native callable method), and open issue on it.

jkotas replied:
(and also delete `ENCODE_METHOD_NATIVE_ENTRY`)

AaronRobinsonMSFT replied:
Oh. I misinterpreted its meaning. Will cleanup.

AaronRobinsonMSFT committed:
Review feedback on crossgen.

jkotas commented on src/coreclr/src/zap/zapinfo.cpp:
```diff
@@ -2268,16 +2268,18 @@ void ZapInfo::getCallInfo(CORINFO_RESOLVED_TOKEN * pResolvedToken,
     }
 #endif
 
+#if defined(TARGET_X86) && defined(TARGET_WINDOWS)
```
Could you please add the same block to src\coreclr\src\tools\crossgen2\ILCompiler.ReadyToRun\JitInterface\CorInfoImpl.ReadyToRun.cs ?

jkotas replied:
(throw new NotImplementedException ...)

jkotas commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -12362,6 +12386,11 @@ CorJitResult CallCompileMethodWithSEHWrapper(EEJitManager *jitMgr,
          }
     }
 
+#if !defined(TARGET_X86) || !defined(TARGET_WINDOWS)
```
The equivalent code in crossgen2 that sets `CORJIT_FLAG_REVERSE_PINVOKE` is not under this conditon. 

It may be best to fail compilation of the native callable methods in both crossgen and crossgen2.

jkotas replied:
(so that we do not have a problem with ABI mismatch once the Windows x86 path is done properly)

jkotas commented on src/coreclr/src/vm/jithelpers.cpp:
```diff
@@ -5041,6 +5041,59 @@ Thread * __stdcall JIT_InitPInvokeFrame(InlinedCallFrame *pFrame, PTR_VOID StubS
+
+    Thread* thread = GetThreadNULLOk();
+    if (thread == NULL)
+        thread = CreateThreadBlockThrow();
```
If the exception thrown from CreateThreadBlockThrow going to show up with the right diagnostic? 

I think we may need

```
    INSTALL_MANAGED_EXCEPTION_DISPATCHER
    INSTALL_UNWIND_AND_CONTINUE_HANDLER
```

around `CreateThreadBlockThrow` to make it work properly.

AaronRobinsonMSFT replied:
Unfortunately can't do that since the `INSTALL_UNWIND_AND_CONTINUE_HANDLER` requires `CURRENT_THREAD` to be non-null and we are trying to address that here. I will update this to handle the case in a fail fast manner though.

kouvel commented:
> @kouvel The diff link doesn't seem to point at an error case or I am just unsure how to interpret github navigation. Can you specifically point at the location in question?

Those comments are at the very bottom of the diff in https://github.com/dotnet/runtime/commit/105d53f2d948b829f24eae4bd18acc16b54e50d5#diff-5d57bafe4ff55258bbe67ca080e6bb71R780 or https://github.com/dotnet/runtime/commit/105d53f2d948b829f24eae4bd18acc16b54e50d5

AaronRobinsonMSFT commented:
>Those comments are at the very bottom of the diff in 105d53f#diff-5d57bafe4ff55258bbe67ca080e6bb71R780 or 105d53f

Yes, I get that. I updated that code. was your comment stating that what I did was okay or something more was needed. In the original comment an "error case" was mentioned that I don't see so I am confused as to what the original comment was alluding at.

kouvel commented:
> In the original comment an "error case" was mentioned that I don't see so I am confused as to what the original comment was alluding at.

Oh ok, yea the link is not working well. With this link https://github.com/dotnet/runtime/commit/c78f4a4288aeee32938e0f96535e52d0978bfb69#diff-b06261f7127af3268df1f78173d20b34R1659 it's on line 1659 on the right side (have to expand the default view to that line and it then should show the highlight).

I think your change would cover this case as well, it would go in with Unknown and use the slow path.

kouvel commented:
Edited above to add file name

AaronRobinsonMSFT committed:
Update CrossGen2 to match CrossGen semantics with respect to NativeCallable.

AaronRobinsonMSFT committed:
Propagate the CallerGCMode in additional scenarios.
Create macro for fail fast create thread.

jkotas commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -2850,6 +2850,7 @@ private uint getJitFlags(ref CORJIT_FLAGS flags, uint sizeInBytes)
             flags.Set(CorJitFlag.CORJIT_FLAG_USE_PINVOKE_HELPERS);
 
             TargetArchitecture targetArchitecture = _compilation.TypeSystemContext.Target.Architecture;
+            TargetOS targetOS = _compilation.TypeSystemContext.Target.OperatingSystem;
```
Nit: Given that this is used just once and only on x86, it does not seem to be worth it to cache it in a local.

AaronRobinsonMSFT committed:
Feedback about local variable.

jkotas commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -1635,6 +1660,14 @@ private void getCallInfo(ref CORINFO_RESOLVED_TOKEN pResolvedToken, CORINFO_RESO
 
             pResult->methodFlags = FilterNamedIntrinsicMethodAttribs(pResult->methodFlags, methodToCall);
 
+            bool _TARGET_X86_ = _compilation.TypeSystemContext.Target.Architecture == TargetArchitecture.X86;
```
```suggestion
```suggestion
```
            var targetDetails = _compilation.TypeSystemContext.Target;
            if (targetDetails.Architecture == TargetArchitecture.X86 
                && targetDetails.OperatingSystem == TargetOS.Windows
                && targetMethod.IsNativeCallable)
{
...
}
```

jkotas replied:
`_TARGET_X86_` as variable name looks very non-CSharpy.

AaronRobinsonMSFT replied:
That is done above, was following the existing style.

jkotas replied:
Fortunately, being deleted in https://github.com/dotnet/runtime/pull/33274/files#diff-030b7d6ddb3dc6ea936a19a6fa55ac59L1475

jkotas commented on src/coreclr/src/vm/jithelpers.cpp:
```diff
@@ -5041,6 +5041,59 @@ Thread * __stdcall JIT_InitPInvokeFrame(InlinedCallFrame *pFrame, PTR_VOID StubS
+
+    Thread* thread = GetThreadNULLOk();
+    if (thread == NULL)
+        CREATETHREAD_IF_NULL_FAILFAST(thread, W("Failed to setup new thread during reverse Platform Invoke"));
```
```suggestion
        CREATETHREAD_IF_NULL_FAILFAST(thread, W("Failed to setup new thread during reverse PInvoke"));
```

We typically use PInvoke or P/Invoke, even in user facing strings. I have searched for "platform invoke" under coreclr and it is used just 3 times in an internal comment. "PInvoke" is used 2000 times, including many occurrences in public facing strings like error messages.

jkotas commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -2881,8 +2882,12 @@ private uint getJitFlags(ref CORJIT_FLAGS flags, uint sizeInBytes)
             }
 
-            if (this.MethodBeingCompiled.IsNativeCallable)
+            if (targetArchitecture == TargetArchitecture.X86
```
This condition is backwards.

It would be best to block R2R for NativeCallable methods on Windows x86 - it will safe us from dealing with the ABI mismatches if/once we get to do the Windows x86 case propertly.

```
if (this.MethodBeingCompiled.IsNativeCallable)
{
    if (targetArchitecture == TargetArchitecture.X86 
        && targetOS == TargetOS.Windows)
    {
        throw new ....;
    }
    flags.Set(CorJitFlag.CORJIT_FLAG_REVERSE_PINVOKE);
}
```

jkotas replied:
It would be with ifdef, like:
```
if (this.MethodBeingCompiled.IsNativeCallable)
{
#if READYTORUN 
    if (targetArchitecture == TargetArchitecture.X86 
        && targetOS == TargetOS.Windows)
    {
        throw new ....;
    }
#endif
    flags.Set(CorJitFlag.CORJIT_FLAG_REVERSE_PINVOKE);
}
```

AaronRobinsonMSFT replied:
Wow. That is embarrassing.

AaronRobinsonMSFT replied:
Wait, why is the `READYTORUN` option relevant here? I wish I understood all these options/configurations the Crossge/ReadyToRun/... scenarios. Every time I read through this code I feel less and less confident that I understand any of it.

jkotas replied:
We try to keep this file shareable with the full AOT compiler in CoreRT. The differences are done via ifdefs currently.

AaronRobinsonMSFT committed:
Style fixes in CrossGen2

AaronRobinsonMSFT committed:
Update error message for FailFast during reverse P/Invoke.

jkotas commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -2882,7 +2882,17 @@ private uint getJitFlags(ref CORJIT_FLAGS flags, uint sizeInBytes)
+                if (targetArchitecture == TargetArchitecture.X86
+                    && _compilation.TypeSystemContext.Target.OperatingSystem == TargetOS.Windows)
+                {
+                    throw new NotSupportedException("ReadyToRun: Methods with NativeCallableAttribute not supported");
```
NotImplementedException may be more appropriate for these, no sure.

AaronRobinsonMSFT replied:
Oh. Did I misread your comment? They were originally `E_NOTIMPL` so that makes sense. I thought you said to make them `NotSupportedException`.

AaronRobinsonMSFT replied:
Yep. I misread your comment.

jkotas commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -2882,7 +2882,17 @@ private uint getJitFlags(ref CORJIT_FLAGS flags, uint sizeInBytes)
+#if READYTORUN
+                if (targetArchitecture == TargetArchitecture.X86
+                    && _compilation.TypeSystemContext.Target.OperatingSystem == TargetOS.Windows)
+                {
```
Could you please add equivalent logic for crossgen1?

E.g. to `ZapInfo::CompileMethod`:

```
#if defined(TARGET_WINDOWS) && (TARGET_X86)
    if (GetCompileInfo()->IsNativeCallableMethod(m_currentMethodHandle))

```

jkotas reviewed (approved):
Just last few nits, but looks great otherwise

AaronRobinsonMSFT committed:
Change CrossGen2 exception type to NotImplementedException.

AaronRobinsonMSFT committed:
Block R2R compilation of NativeCallableAttribute methods for crossgen.exe.

MichalStrehovsky commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -1635,6 +1660,14 @@ private void getCallInfo(ref CORINFO_RESOLVED_TOKEN pResolvedToken, CORINFO_RESO
+                && targetDetails.OperatingSystem == TargetOS.Windows
+                && targetMethod.IsNativeCallable)
+            {
+                throw new NotImplementedException("ReadyToRun: References to methods with NativeCallableAttribute not implemented");
```
AFAIK this is going to tear down the compilation process. Throwing `RequiresRuntimeJitException` would be more appropriate because that one gets handled.

(Same comment for the NotImplement in CorInfoImpl.cs)

AaronRobinsonMSFT committed:
Throw RequiresRuntimeJitException instead of NotImplementedException in Crossgen2.

AaronRobinsonMSFT commented:
The Linux x64 Debug and Checked versions passed prior to the [exception change in CrossGen2](https://github.com/dotnet/runtime/pull/33005/commits/48758d4218de1e9cb624df12689ec4b69039889f) so assuming they are not related to any changes here.
